enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/config/ConfigBuilderProvider.java

```diff
@@ -193,16 +209,13 @@ public static ConfigBuilder getBuilder(final GlassFishServer server) {
                     "GlassFish server entity shall not be null");
         }
         ConfigBuilder builder;
-        synchronized (builders) {
-            builder = builders.get(server);
-            if (builder != null) {
-                return builder;
-            }
-            String serverHome = server.getServerHome();
-            builder = new ConfigBuilder(config, serverHome, serverHome, serverHome);
-            builders.put(server, builder);
+        builder = builders.computeIfPresent(server, (key, value) -> value);
+        if (builder != null) {
+            return builder;
         }
-        return builder;
+        String serverHome = server.getServerHome();
+        return builders.computeIfAbsent(server, key -> 
+                new ConfigBuilder(config, serverHome, serverHome, serverHome));
```

Comment:
something is wrong here. Please take another look.
`computeIfPresent(server, (key, value) -> value)` is essentially `get()` and I believe this was meant to be removed?

Reply:
The `computeIfPresent` implementation from `ConcurrentHashMap` synchronize at a node level (`Node<K,V>`).  I use `computeIfPresent` to circumvent the need of synchronize the entire map. It does work like a `get()` (`get()` does not synchronize) and it does not have the `@Deprecated` as of now.

Reply:
the old version:
```java
synchronized {
    // checked if `server` was in the map, if it was, it returned it.
    // if it wasn't in the map, it created a new ConfigBuilder, registered it in the map and returned it
}
```
this is the same as:
```java
        String serverHome = server.getServerHome();
        return builders.computeIfAbsent(server, key -> 
                new ConfigBuilder(config, serverHome, serverHome, serverHome));
```
(If i am missing something, please explain)

Reply:
Both versions will compute the same results. But the old version does not allow concurrent modifications to the map, the new version will allow concurrent modifications to the map.
Old version:
```java
synchronized (entireMap) {
    // checked if `server` key was in the map, if it was, it returned it's value.
    // if it wasn't in the map, it created a new ConfigBuilder, registered it in the map and returned it
}
```
New version restrict modifications at a node level e.g. `Node[K1,V1]` until it finish, but allows  concurrent modifications to every other node e.g. `Node[K2,V2]`:
```java
synchronized (Node[K1,V1]) { 
    // Thread 1
    // No one can modify this node until I finish
}
synchronized (Node[K2,V2]) {
    // Thread 2
    // Modifying this node at the same time that Thread 1 is working on Node[K1,V1]
}
```

Reply:
the the task of the method, which is to return the value of the map or compute it if it isn't there yet, can be written with one operation on the map:
```java
    public static ConfigBuilder getBuilder(final GlassFishServer server) {
        if (server == null) {
            throw new ServerConfigException(
                    "GlassFish server entity shall not be null");
        }
        String serverHome = server.getServerHome();
        return builders.computeIfAbsent(server, key -> 
                new ConfigBuilder(config, serverHome, serverHome, serverHome));
    }

```

Reply:
You are right!, everything will be done in one atomic and concurrent statement, great work @mbien.
I will re-test everything and push the update.

Reply:
figured out how to create a jackpot rule for this, nb can refactor it now with one click:
https://github.com/mbien/jackpot-inspections/commit/8eadafc5c45c3a1898ebadaa16455c5019537c9c